lib/galaxy/tools/__init__.py

```diff
@@ -3000,6 +3046,170 @@ def job_failed(self, job_wrapper, message, exception=False):
+                    v["nameext"] = "None"
+                    v["nameroot"] = directory_name
+                    v["basename"] = directory_name
+
+        cwl_job_proxy = self._cwl_tool_proxy.job_proxy(
+            input_json,
+            output_dict,
+            local_working_directory,
+        )
+        cwl_command_line = cwl_job_proxy.command_line
+        cwl_stdin = cwl_job_proxy.stdin
+        cwl_stdout = cwl_job_proxy.stdout
+        cwl_stderr = cwl_job_proxy.stderr
+        env = cwl_job_proxy.environment
+
+        def needs_shell_quoting_hack(arg):
+            if arg == "$GALAXY_SLOTS":
+                return False
+            else:
+                return needs_shell_quoting(arg)
+
+        command_line = " ".join(
+            shellescape.quote(arg) if needs_shell_quoting_hack(arg) else arg for arg in cwl_command_line
+        )
+        if cwl_stdin:
+            command_line += ' < "' + cwl_stdin + '"'
+        if cwl_stdout:
+            command_line += ' > "' + cwl_stdout + '"'
+        if cwl_stderr:
+            command_line += ' 2> "' + cwl_stderr + '"'
+        cwl_job_state = {
+            "args": cwl_command_line,
+            "stdin": cwl_stdin,
+            "stdout": cwl_stdout,
+            "stderr": cwl_stderr,
+            "env": env,
+        }
+        tool_working_directory = os.path.join(local_working_directory, "working")
+        # Move to prepare...
+        safe_makedirs(tool_working_directory)
+        cwl_job_proxy.stage_files()
+
+        cwl_job_proxy.rewrite_inputs_for_staging()
+        # Write representation to disk that can be reloaded at runtime
+        # and outputs collected before Galaxy metadata is gathered.
+        cwl_job_proxy.save_job()
+
+        param_dict["__cwl_command"] = command_line
+        param_dict["__cwl_command_state"] = cwl_job_state
+        param_dict["__cwl_command_version"] = 1
+        log.info("CwlTool.exec_before_job() generated command_line %s", command_line)
+
+    def parse(self, tool_source, **kwds):
+        super().parse(tool_source, **kwds)
+        cwl_tool_proxy = getattr(tool_source, "tool_proxy", None)
+        if cwl_tool_proxy is None:
+            raise Exception("parse() called on tool source not defining a proxy object to underlying CWL tool.")
+        self._cwl_tool_proxy = cwl_tool_proxy
+
+    def param_dict_to_cwl_inputs(self, param_dict, local_working_directory):
+        """Map Galaxy API inputs description to a CWL job json."""
+        raise NotImplementedError()
+
+
+class GalacticCwlTool(CwlCommandBindingTool):
+    """A CWL tool with a gx:Interface defined so Galaxy tool state can be used."""
+
+    tool_type = "galactic_cwl"
+
+    def param_dict_to_cwl_inputs(self, param_dict, local_working_directory):
+        from galaxy.tool_util.cwl.representation import galactic_flavored_to_cwl_job
+
+        input_json = galactic_flavored_to_cwl_job(self, param_dict, local_working_directory)
+        return input_json
+
+
```

Comment:
Do we still want to add `GalacticCwlTool` or should we drop this?

Reply:
Do we have to make a decision here ? I think it's a good idea, but if we remove this we should put it into a separate commit (if it isn't already)

Reply:
Do you mean remove it in a separate commit?